activation/validation.go

```diff
@@ -263,16 +259,16 @@ func (*Validator) NIPostChallenge(challenge *types.NIPostChallenge, atxs atxProv
 		)
 	}
 
-	if prevATX.PublishEpoch >= challenge.PublishEpoch {
+	if prevATX.PublishEpoch >= challenge.Publish() {
 		return fmt.Errorf(
 			"prevAtx epoch (%d) isn't older than current atx epoch (%d)",
-			prevATX.PublishEpoch, challenge.PublishEpoch,
+			prevATX.PublishEpoch, challenge.Publish(),
 		)
 	}
 
-	if prevATX.Sequence+1 != challenge.Sequence {
+	if seq := challenge.MaybeSequence(); seq != nil && prevATX.Sequence+1 != *seq {
 		return fmt.Errorf(
-			"sequence number (%d) is not one more than the prev one (%d)", challenge.Sequence, prevATX.Sequence)
+			"sequence number (%d) is not one more than the prev one (%d)", *seq, prevATX.Sequence)
```

Comment:
OK I think I understand now why the abstraction of the challenge through the interface. I don't think it is a good idea to have the same validation method for both ATX v1 and ATX v2 and hiding the actual ATX version from the validator.

This might look initially like a good way to go because we save on having 2 validators, but it will lead to code like this sequence check where now a function that shouldn't be aware which ATX version it is validating now kinda needs to be.

Additionally `Sequence` is not optional in a v1 ATX which is not clear from this code - I have to look at the implementation details of `MaybeSequence` to see that a V1 ATX will always return a sequence number while a V2 ATX will (probably) always return `nil`. It gives the impression that "if it has one validate it" - which is not the case. v1 must have one, v2 must/can not.

Every extension/fix/improvement to validation with this generic approach also has the potential to break an already working validation path for older ATX versions. Ideally we would tread the V1 validator like legacy code; we never touch it not even to add new features or improve performance unless we know there is an issue that specifically happens in that part of the codebase.

With a generic approach eventually this method will grow in complexity just because newer ATXs need to be validated, potentially breaking validation of older ATXs.

Reply:
removed the interface and changed the methods to take the `wire.NipostChallengeV1`.